prov/lnx/src/lnx_cq.c

```diff
@@ -0,0 +1,241 @@
+	.writeerr = lnx_peer_cq_writeerr,
+};
+
+static struct fi_ops lnx_cq_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = lnx_cq_close,
+	.bind = fi_no_bind,
+	.control = ofi_cq_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static void lnx_cq_progress(struct util_cq *cq)
+{
+	struct lnx_cq *lnx_cq;
+	struct local_prov_ep *ep;
+	struct local_prov *entry;
+	struct dlist_entry *prov_table;
+
+	lnx_cq = container_of(cq, struct lnx_cq, util_cq);
+	prov_table = &lnx_cq->lnx_domain->ld_fabric->local_prov_table;
+/*
+	entry = lnx_cq->lnx_domain->ld_fabric->shm_prov;
+	dlist_foreach_container(&entry->lpv_prov_eps,
+				struct local_prov_ep, ep, entry)
+		fi_cq_read(ep->lpe_cq.lpc_core_cq, NULL, 0);
+
+	return;
+*/
+	/* Kick the core provider endpoints to progress */
+	dlist_foreach_container(prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+					struct local_prov_ep, ep, entry)
+			fi_cq_read(ep->lpe_cq.lpc_core_cq, NULL, 0);
+	}
+}
+
+static int lnx_cq_open_core_prov(struct lnx_cq *cq, struct fi_cq_attr *attr)
+{
+	int rc;
+	struct local_prov_ep *ep;
+	struct local_prov *entry;
+	struct dlist_entry *prov_table =
+		&cq->lnx_domain->ld_fabric->local_prov_table;
+
+	/* tell the core providers to import my CQ */
+	attr->flags |= FI_PEER;
+
+	/* create all the core provider endpoints */
+	dlist_foreach_container(prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+					struct local_prov_ep, ep, entry) {
+			struct fid_cq *core_cq;
+			struct fi_peer_cq_context cq_ctxt;
+
+			ep->lpe_cq.lpc_shared_cq = cq;
+			ep->lpe_cq.lpc_cq.owner_ops = &lnx_cq_write;
+
+			cq_ctxt.size = sizeof(cq_ctxt);
+			cq_ctxt.cq = &ep->lpe_cq.lpc_cq;
+
+			/* pass my CQ into the open and get back the core's cq */
+			rc = fi_cq_open(ep->lpe_domain, attr, &core_cq, &cq_ctxt);
+			if (rc)
+				return rc;
+
+			/* before the fi_cq_open() returns the core provider should
+			 * have called fi_export_fid() and got a pointer to the peer
+			 * CQ which we have allocated for this core provider
+			 */
+
+			ep->lpe_cq.lpc_core_cq = core_cq;
+		}
+	}
+
+	return 0;
+}
+
+int lnx_cq_open(struct fid_domain *domain, struct fi_cq_attr *attr,
+		struct fid_cq **cq_fid, void *context)
+{
+	struct lnx_cq *lnx_cq;
+	struct lnx_domain *lnx_dom;
+	int rc;
+
+	lnx_cq = calloc(1, sizeof(*lnx_cq));
+	if (!lnx_cq)
+		return -FI_ENOMEM;
+
+	/* this is going to be a standard CQ from the read side. From the
+	 * write side, it'll use the peer_cq callbacks to write 
+	 */
+	rc = ofi_cq_init(&lnx_prov, domain, attr, &lnx_cq->util_cq,
+			 &lnx_cq_progress, context);
+	if (rc)
+		goto free;
+
+	lnx_dom = container_of(domain, struct lnx_domain,
+			       ld_domain.domain_fid);
+
+	lnx_cq->lnx_domain = lnx_dom;
+	lnx_cq->util_cq.cq_fid.fid.ops = &lnx_cq_fi_ops;
+	(*cq_fid) = &lnx_cq->util_cq.cq_fid;
+
+	/* open core CQs and tell them to import my CQ */
+	rc = lnx_cq_open_core_prov(lnx_cq, attr);
```

Comment:
since *cq_fid is set before this function is called, then the application should clean it up by calling fi_close(). When that gets processed by LNX it'll close all the underlying core providers.

Reply:
The validity of *cq_fid should be determine by return code and not the setting of *cq_fid. I would not assume that all apps check *cq_fid being non-NULL if RC != FI_SUCCESS.